sphinx/builders/linkcheck.py

```diff
@@ -437,9 +449,30 @@ def _check_uri(self, uri: str, hyperlink: Hyperlink) -> tuple[str, str, int]:
             except HTTPError as err:
                 error_message = str(err)
 
-                # Unauthorised: the reference probably exists
+                # Unauthorized: the client did not provide required credentials
                 if status_code == 401:
-                    return 'working', 'unauthorized', 0
+                    if self._allow_unauthorized:
+                        deprecation_msg = (
+                            "\n---\n"
+                            "The linkcheck builder encountered an HTTP 401 "
+                            f"(unauthorized) response for the URI {uri}, and "
```

Comment:
Adding the `uri` here could make this spammy for some deployments, if they are encountering large numbers of HTTP 401s.  I would guess (?) that the typical case is likely a few/rare cases, though - in which case the message is more actionable.  I could be convinced to remove the dynamic URI, though.

Reply:
(to explain that: the [default Python warning logic](https://docs.python.org/3/library/warnings.html#the-warnings-filter) is to print each _unique_ warning -- based on the message, module and line number -- once.   having a dynamic URI means that this message would appear once per URI)

Reply:
Ok, here should be the final draft (dw I also want to be sure about what we are doing).

----

Currently, a broken link is reported either with `logger.info` in red or with `logger.warn` when `-W` is used. In the future, we want all 401 to be "broken". To make the transition as smooth as possible, I suggest the following:

#### Case 1 

If the configuration value is not changed, 401 are treated as working and reported by `logger.info` in green but this message will contain the link and the fact that it was not authorized.

However, a depreciation warning needs to be emitted. Since the user didn't touch the configuration, this depr. warn.  needs to be emitted when encountering  the first 401. Users will know which links to change since (working) 401 are actually printed.

Note that it doesn't matter whether `-W` is set since 401 are considered as working anyway.

#### Case 2

If the configuration value is changed by the user, then 401 are considered as broken. Since they opted in for this feature in 7.3.x, this means that they want these 401 to be turned into real warnings in 8.x. Note that treating warnings as errors is not problematic since users asked for 401 to be explicitly broken. In particular, with `-W`, they expect them to be warnings (hence errors), and the build to fail. Without `-W`, they still expect a red message. So we simply need to report 401 as broken.

In addition, since they touched the configuration value, they will be warned by a deprecation warning in the constructor. Again, this warning won't cause any failure with `-W`.

Reply:
Both of those descriptions match my understanding of the intended behaviour, with one small detail/confirmation to check.  Thank you very much @picnixz for writing this up :+1: 

My detail check is in response to this statement in case 1:

> this depr. warn. needs to be emitted when encountering the first 401.

I think this implies that we should _not_ include the `uri` inside the deprecation warning logged in this scenario.

Reply:
Indeed. Because they are already printed separately so I don't think it's good to rewrite them once again.